File: app/src/main/java/com/vondear/tools/adapter/AdapterLeftMenu.java
package com.vondear.tools.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.vondear.tools.R;
import com.vondear.tools.model.ModelDishMenu;

import java.util.ArrayList;
import java.util.List;

/**
 *
 * @author vondear
 * @date 16-11-10
 */
public class AdapterLeftMenu extends RecyclerView.Adapter {
    private Context mContext;
    private ArrayList<ModelDishMenu> mMenuList;
    private int mSelectedNum;
    private List<onItemSelectedListener> mSelectedListenerList;

    public interface onItemSelectedListener{
        public void onLeftItemSelected(int postion, ModelDishMenu menu);
    }

    public void addItemSelectedListener(onItemSelectedListener listener){
        if(mSelectedListenerList!=null)
            mSelectedListenerList.add(listener);
    }

    public void removeItemSelectedListener(onItemSelectedListener listener){
        if(mSelectedListenerList!=null && !mSelectedListenerList.isEmpty())
            mSelectedListenerList.remove(listener);
    }

    public AdapterLeftMenu(Context mContext, ArrayList<ModelDishMenu> mMenuList){
        this.mContext = mContext;
        this.mMenuList = mMenuList;
        this.mSelectedNum = -1;
        this.mSelectedListenerList = new ArrayList<>();
        if(mMenuList.size()>0)
            mSelectedNum = 0;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.left_menu_item, parent, false);
        LeftMenuViewHolder viewHolder = new LeftMenuViewHolder(view);
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        ModelDishMenu modelDishMenu = mMenuList.get(position);
        LeftMenuViewHolder viewHolder = (LeftMenuViewHolder)holder;
        viewHolder.menuName.setText(modelDishMenu.getMenuName());
        if(mSelectedNum==position){
            viewHolder.menuLayout.setSelected(true);
        }else{
            viewHolder.menuLayout.setSelected(false);
        }
    }

    @Override
    public int getItemCount() {
        return mMenuList.size();
    }

    public void setSelectedNum(int selectedNum) {
        if(selectedNum<getItemCount() && selectedNum>=0 ) {
            this.mSelectedNum = selectedNum;
            notifyDataSetChanged();
        }
    }

    public int getSelectedNum() {
        return mSelectedNum;
    }

    private class LeftMenuViewHolder extends RecyclerView.ViewHolder{

        TextView menuName;
        LinearLayout menuLayout;

        public LeftMenuViewHolder(final View itemView) {
            super(itemView);
            menuName = (TextView)itemView.findViewById(R.id.left_menu_textview);
            menuLayout = (LinearLayout)itemView.findViewById(R.id.left_menu_item);
            menuLayout.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    int clickPosition = getAdapterPosition();
//                    setSelectedNum(clickPosition);
                    notifyItemSelected(clickPosition);
                }
            });
        }
    }

    private void notifyItemSelected(int position) {
        if(mSelectedListenerList!=null && !mSelectedListenerList.isEmpty()){
            for(onItemSelectedListener listener:mSelectedListenerList){
                listener.onLeftItemSelected(position,mMenuList.get(position));
            }
        }
    }
}

File: RxDemo/src/main/java/com/tamsiree/rxdemo/adapter/AdapterLeftMenu.kt
package com.tamsiree.rxdemo.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.tamsiree.rxdemo.R
import com.tamsiree.rxdemo.model.ModelDishMenu
import java.util.*

/**
 *
 * @author tamsiree
 * @date 16-11-10
 */
class AdapterLeftMenu(private val mContext: Context, private val mMenuList: ArrayList<ModelDishMenu>) : RecyclerView.Adapter<RecyclerView.ViewHolder?>() {
    private var mSelectedNum: Int
    private val mSelectedListenerList: MutableList<onItemSelectedListener>?

    interface onItemSelectedListener {
        fun onLeftItemSelected(postion: Int, menu: ModelDishMenu)
    }

    fun addItemSelectedListener(listener: onItemSelectedListener) {
        mSelectedListenerList?.add(listener)
    }

    fun removeItemSelectedListener(listener: onItemSelectedListener) {
        if (mSelectedListenerList != null && mSelectedListenerList.isNotEmpty()) mSelectedListenerList.remove(listener)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.left_menu_item, parent, false)
        return LeftMenuViewHolder(view)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val modelDishMenu = mMenuList[position]
        val viewHolder = holder as LeftMenuViewHolder?
        viewHolder!!.menuName.text = modelDishMenu.menuName
        viewHolder.menuLayout.isSelected = mSelectedNum == position
    }

    override fun getItemCount(): Int {
        return mMenuList.size
    }

    var selectedNum: Int
        get() = mSelectedNum
        set(selectedNum) {
            if (selectedNum in 0 until itemCount) {
                mSelectedNum = selectedNum
                notifyDataSetChanged()
            }
        }

    private inner class LeftMenuViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var menuName: TextView = itemView.findViewById(R.id.left_menu_textview)
        var menuLayout: LinearLayout = itemView.findViewById(R.id.left_menu_item)

        init {
            menuLayout.setOnClickListener {
                val clickPosition = adapterPosition
                //                    setSelectedNum(clickPosition);
                notifyItemSelected(clickPosition)
            }
        }
    }

    private fun notifyItemSelected(position: Int) {
        if (mSelectedListenerList != null && mSelectedListenerList.isNotEmpty()) {
            for (listener in mSelectedListenerList) {
                listener.onLeftItemSelected(position, mMenuList[position])
            }
        }
    }

    init {
        mSelectedNum = -1
        mSelectedListenerList = ArrayList()
        if (mMenuList.size > 0) mSelectedNum = 0
    }
}